File: app/src/main/java/mrkm4ntr/twitterclient/data/TwitterProvider.java
package mrkm4ntr.twitterclient.data;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.Nullable;

public class TwitterProvider extends ContentProvider {

    private static final UriMatcher URI_MATCHER = buildUriMatcher();
    private TwitterDbHelper mOpenHelper;

    private static final int STATUS = 100;
    private static final int STATUS_WITH_ID = 101;
    private static final int USER = 200;
    private static final int USER_WITH_ID = 201;

    private static UriMatcher buildUriMatcher() {
        UriMatcher matcher = new UriMatcher(UriMatcher.NO_MATCH);
        String authority = TwitterContract.CONTENT_AUTHORITY;
        matcher.addURI(authority, TwitterContract.PATH_STATUS, STATUS);
        matcher.addURI(authority, TwitterContract.PATH_STATUS + "/*", STATUS_WITH_ID);
        // TODO implement
        return matcher;
    }

    @Override
    public boolean onCreate() {
        mOpenHelper = new TwitterDbHelper(getContext());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(Uri uri, String[] projection, String selection, String[] selectionArgs,
                        String sortOrder) {
        Cursor cursor;
        switch (URI_MATCHER.match(uri)) {
            case STATUS: {
                cursor = mOpenHelper.getReadableDatabase().query(
                        TwitterContract.StatusEntry.TABLE_NAME, projection, selection,
                        selectionArgs, null, null, sortOrder);
                break;
            }
            case STATUS_WITH_ID: {
                cursor = mOpenHelper.getReadableDatabase().query(
                        TwitterContract.StatusEntry.TABLE_NAME, projection,
                        TwitterContract.StatusEntry._ID + " = ? ",
                        new String[]{ uri.getPathSegments().get(1) }, null, null, sortOrder);
                break;
            }
            default:
                throw new UnsupportedOperationException("Unknown uri:" + uri);
        }
        cursor.setNotificationUri(getContext().getContentResolver(), uri);
        return cursor;
    }

    @Nullable
    @Override
    public String getType(Uri uri) {
        switch (URI_MATCHER.match(uri)) {
            case STATUS:
                return TwitterContract.StatusEntry.CONTENT_TYPE;
            case STATUS_WITH_ID:
                return TwitterContract.StatusEntry.CONTENT_ITEM_TYPE;
            case USER_WITH_ID:
                return TwitterContract.UserEntry.CONTENT_ITEM_TYPE;
            // TODO implement
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
    }

    @Nullable
    @Override
    public Uri insert(Uri uri, ContentValues contentValues) {
        SQLiteDatabase db = mOpenHelper.getWritableDatabase();
        int match = URI_MATCHER.match(uri);
        Uri returnUri;

        switch (match) {
            case STATUS: {
                long _id = db.insert(TwitterContract.StatusEntry.TABLE_NAME, null, contentValues);
                if (_id > 0) {
                    returnUri = TwitterContract.StatusEntry.buildStatusUri(_id);
                } else {
                    throw new SQLException("Failed to insert row into " + uri);
                }
                break;
            }
            case USER: {
                long _id = db.insert(TwitterContract.UserEntry.TABLE_NAME, null, contentValues);
                if (_id > 0) {
                    returnUri = TwitterContract.UserEntry.buildUserUri(_id);
                } else {
                    throw new SQLException("Failed to insert row into " + uri);
                }
                break;
            }
            default:
                throw  new UnsupportedOperationException("Unknown uri:" + uri);
        }
        getContext().getContentResolver().notifyChange(uri, null);
        return returnUri;
    }

    @Override
    public int bulkInsert(Uri uri, ContentValues[] values) {
        final SQLiteDatabase db = mOpenHelper.getWritableDatabase();
        final int match = URI_MATCHER.match(uri);
        switch (match) {
            case USER:
                db.beginTransaction();
                int returnCount = 0;
                try {
                    for (ContentValues value : values) {
                        long id = db.insert(TwitterContract.UserEntry.TABLE_NAME, null, value);
                        if (id != -1) {
                            returnCount++;
                        }
                    }
                    db.setTransactionSuccessful();
                } finally {
                    db.endTransaction();
                }
                getContext().getContentResolver().notifyChange(uri, null);
                return returnCount;
            default:
                return super.bulkInsert(uri, values);
        }
    }

    @Override
    public int delete(Uri uri, String s, String[] strings) {
        return 0;
    }

    @Override
    public int update(Uri uri, ContentValues contentValues, String s, String[] strings) {
        return 0;
    }
}

File: app/src/main/kotlin/mrkm4ntr/twitterclient/data/TwitterProvider.kt
package mrkm4ntr.twitterclient.data

import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.net.Uri

class TwitterProvider : ContentProvider() {
    private var mOpenHelper: TwitterDbHelper? = null

    override fun onCreate(): Boolean {
        mOpenHelper = TwitterDbHelper(context)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?,
                       selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        val cursor = when (uriMather.match(uri)) {
            STATUS -> {
                mOpenHelper!!.readableDatabase.query(
                        TwitterContract.StatusEntry.TABLE_NAME, projection, selection,
                        selectionArgs, null, null, sortOrder)
            }
            STATUS_WITH_ID -> {
                mOpenHelper!!.readableDatabase.query(
                        TwitterContract.StatusEntry.TABLE_NAME, projection,
                        TwitterContract.StatusEntry._ID + " = ? ",
                        arrayOf<String>(uri.pathSegments[1]), null, null, sortOrder)
            }
            else -> throw UnsupportedOperationException("Unknown uri:" + uri)
        }
        cursor.setNotificationUri(context!!.contentResolver, uri)
        return cursor
    }

    override fun getType(uri: Uri): String? {
        when (uriMather.match(uri)) {
            STATUS -> return TwitterContract.StatusEntry.CONTENT_TYPE
            STATUS_WITH_ID -> return TwitterContract.StatusEntry.CONTENT_ITEM_TYPE
            else -> throw UnsupportedOperationException("Unknown uri: " + uri)
        }
    }

    override fun insert(uri: Uri, contentValues: ContentValues?): Uri? {
        val db = mOpenHelper!!.writableDatabase

        val returnUri = when (uriMather.match(uri)) {
            STATUS -> {
                val _id = db.insert(TwitterContract.StatusEntry.TABLE_NAME, null, contentValues)
                if (_id > 0) {
                    TwitterContract.StatusEntry.buildStatusUri(_id)
                } else {
                    throw SQLException("Failed to insert row into " + uri)
                }
            }
            else -> throw UnsupportedOperationException("Unknown uri:" + uri)
        }
        context!!.contentResolver.notifyChange(uri, null)
        return returnUri
    }

    override fun bulkInsert(uri: Uri, values: Array<ContentValues>): Int {
        val db = mOpenHelper!!.writableDatabase

        when (uriMather.match(uri)) {
            STATUS -> {
                db.beginTransaction()
                val returnCount = try {
                    val count = values.count {
                        db.insert(TwitterContract.StatusEntry.TABLE_NAME, null, it) != -1L
                    }
                    db.setTransactionSuccessful()
                    count
                } finally {
                    db.endTransaction()
                }
                context!!.contentResolver.notifyChange(uri, null)
                return returnCount
            }
            else -> return super.bulkInsert(uri, values)
        }
    }

    override fun delete(uri: Uri, s: String?, strings: Array<String>?): Int {
        return 0
    }

    override fun update(uri: Uri, contentValues: ContentValues?, s: String?, strings: Array<String>?): Int {
        return 0
    }

    companion object {

        private val uriMather = UriMatcher(UriMatcher.NO_MATCH).apply {
            val authority = TwitterContract.CONTENT_AUTHORITY
            addURI(authority, TwitterContract.PATH_STATUS, STATUS)
            addURI(authority, TwitterContract.PATH_STATUS + "/*", STATUS_WITH_ID)
        }

        private val STATUS = 100
        private val STATUS_WITH_ID = 101

    }
}
